src/inference/InterfaceExtensions.jl

```diff
@@ -0,0 +1,334 @@
+module InterfaceExtensions
+
+using ..Turing
+using ..Interface
+import MCMCChains: Chains
+
+import ..Interface: AbstractTransition, sample, step!, sample_init!,
+    transitions_init, sample_end!, AbstractSampler, transition_type
+import ..Turing: Model, Sampler
+import ..RandomVariables: islinked, invlink!, getlogp, tonamedtuple
+import ..Inference: InferenceAlgorithm, ParticleInference
+
+export sample,
+       sample_init!,
+       sample_end!,
+       set_resume!,
+       resume,
+       save,
+       transition_type,
+       transitions_init,
+       transition,
+       Transition,
+       ParticleTransition,
+       Chains
+
+# Internal variable names for MCMCChains.
+const TURING_INTERNAL_VARS =
+    Dict(:internals => ["elapsed", "eval_num", "lf_eps", "lp", "weight", "le",
+                        "acceptance_rate", "hamiltonian_energy", "is_accept", 
+                        "log_density", "n_steps", "numerical_error", "step_size",
+                        "tree_depth"])
+
+####################
+# Transition Types #
+####################
+
+# Used by all non-particle samplers.
+struct Transition{T} <: AbstractTransition
+    θ  :: T
+    lp :: Float64
+end
+
+function transition(spl::Sampler)
+    theta = tonamedtuple(spl.state.vi)
+    lp = getlogp(spl.state.vi)
+    return Transition{typeof(theta)}(theta, lp)
+end
+
+function transition(spl::Sampler, nt::NamedTuple)
+    theta = merge(tonamedtuple(spl.state.vi), nt)
+    lp = getlogp(spl.state.vi)
+    return Transition{typeof(theta)}(theta, lp)
+end
+
+function additional_parameters(::Type{Transition})
+    return [:lp]
+end
+
+# used by PG and SMC
+struct ParticleTransition{T} <: AbstractTransition
```

Comment:
I'm not sure about that. Maybe it would be better to have a `Particle` type that contains a transition as field?

Reply:
That would defeat the purpose of the interface methods. You call `step!` on some sampler, and it returns an `AbstractTransition` object that contains whatever information about the sample you want to make it into a `Chains` object. Each sampler type should have it's own way of bundling up samples -- for PG and SMC, they just extract some parameterization from a particle and stick it in here.

Reply:
Hm, I think what confuses me is the terminology.

Reply:
Ah, I see. `ParticleTransition` is just a way to bundle up a specific sample, and this one just happens to be a specific implementation for the particle samplers. I'm not really clear on why we use the word "Transition" here and not something else. I think we're following how AdvancedHMC is doing everything.